### ISSUE TITLE
fsDateConverter cannot serialize all datetime strings

### ISSUE THREAD

**Issue by Kasperki** [jacobdufault/fullserializer#77]
In fsDateConverter.cs
TryDeserialize DateTime.TryParse() can't parse this string: 2016-01-22T12:06:57.503005Z
but Convert.ToDateTime() can.

**Comment by jacobdufault** [comment]
I'm not super familiar with time parsing. Do you know what we need to pass to `DateTime.TryParse` to make it work?

If necessary, I can wrap the `Convert.ToDateTime` call in a try/catch exception block, but FS tries really hard to not throw exceptions since it causes a big perf hit on AOT platforms (or it did for iOS - I need to reeval for il2cpp).

**Comment by Kasperki** [comment]
@jacobdufault no I don't know what we need to pass to DateTime.TryParse to get it working, I tried somethings but did not get it working. My quick fix solution was still to use TryParse and if that can't parse it then fallback and try Convert.ToDateTime. I'm using this for webgl, because it does not support reflection, what many serializers use.

**Comment by paul-westpierstudio** [comment]
Hi @jacobdufault and @Kasperki

I encountered this same issue recently, and discovered that the serializer (Newtonsoft, in this case) was using a format that produces a slightly more optimal string output for date/times, by truncating trailing zeroes from the timestamp part.

In fsDateConverter, DateTime.TryParse uses DateTimeStyles.RoundtripKind, which requires the format to be "yyyy'-'MM'-'dd'T'HH':'mm':'ss'.'fffffffK" (note lowercase "f").

In my case, the Newtonsoft serializer uses "yyyy'-'MM'-'dd'T'HH':'mm':'ss.FFFFFFFK" (note uppercase "F").  This truncates trailing zeroes.  I don't know if this is still technically ISO 8601 compliant.

This looks to be the case as in your example above: "2016-01-22T12:06:57.503005Z"

The fractional seconds part should have exactly 7 digits, in order for DateTime.TryParse to work: the example above only has 6 digits, so DateTime.TryParse fails, ie. this would work fine: "2016-01-22T12:06:57.5030050Z"

However, I think the fix pushed in commit 5b9f8f3 might need some further thought: the output of Convert.ToDateTime is a local date/time, even when the input string represents a UTC date/time, so I would say that even thought the deserializer is successful, the output does not faithfully represents the input.

Also, I believe that Convert.ToDateTime is culture-specific, so the interpretation of the input string is dependent on the executing thread's current culture; the conversion may fail altogether, or return a very different and possibly erroneous result (eg. day/month, vs. month/day).

I haven't tested the case where the input string represents a time in a different timezone;  and whether Convert.ToDateTime always produces a local time, but I would suggest at the very least that a warning message be added to the deserializer result, when DateTime.TryParse and the fallback is used.

In my case, I was able to adjust the Newtonsoft SerializerSettings to use a DateFormatString of "o", which is Microsoft's full round-trip format, with no truncation - this ensured that DateTime.TryParse always worked.

See https://docs.microsoft.com/en-us/dotnet/standard/base-types/standard-date-and-time-format-strings#Roundtrip

Hope that helps.

Thanks
Paul
